NativeAOT fixes around data flow analysis

vitek-karas opened the pull request:
Removes the `Origin` class as it doesn't seem to serve any purpose anymore. The code used it to basically just only call `GetDisplayName` on members and pass that as the `reason` to dependency analysis. Replaced with passing just string directly as the `reason`.

Fixes bugs in warning subsystem when the warning is generated from an origin which is a module or assembly. This would previously crash the compiler. There are basically no tests for this right now - some of the tests in linker which hit this can't be easily ported yet due to NativeAOT not producing "reflection access to DAM" warnings. I have these mostly working locally but enabling the warnings fails some tests outside of the linker ones, so it needs more work.

Fixes analysis bugs when methods or fields are accessed via `ldtoken` or `ldftn` instructions. In such cases we still need to process them for RUC and DAM based warnings and so on. There are only a few tests for this in this change - more will come when we port the `AnnotatedMembersAccessedViaReflection` from linker, but those all require to enable the "reflection access to DAM" warnings as mentioned above.
* This required addition of a new TrimAnalysisPattern since the processing happens in reflection scanning and all warnings/marking hapening from there must go through the patterns.

Changes `DynamicDependencyAttribute` processing to happen in a separate dependency node. Previously if there were warnings generated due to the `DynamicDependencyAttribute` the compiler may report them multiple times (if it hit the same member twice during the rest of the analysis). To avoid this, the processing is postponed by adding a new dependency node - which then acts as the unification mechanism and thus warnings are produced just once.

vitek-karas committed:
First

vitek-karas committed:
Remove origin progress

vitek-karas committed:
Progress

vitek-karas committed:
Test tweaks

vitek-karas committed:
Progress

vitek-karas committed:
Make it all work

vitek-karas committed:
Fix ldtoken

vitek-karas committed:
Fix usings

vitek-karas committed:
Fix a test

vitek-karas committed:
Revert enable of reflection access warnings for now

vitek-karas committed:
Remove unnecessary change to the project file

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:
```diff
@@ -954,6 +957,9 @@ private void ImportLdToken(int token)
             else
             {
                 Debug.Assert(obj is FieldDesc);
+                var field = (FieldDesc)obj;
```
The assert is now of little value.

```suggestion
                var field = (FieldDesc)obj;
```

MichalStrehovsky reviewed (approved):
Thank you!

vitek-karas committed:
PR Feedback and fix failures

The failures are due to Canon instantiations which are not expected in some of the reflection specific parts of the code. Since the change calls the GetDependenciesDueToAccess on new things, it hits this as well. Fixed by checking for Canon instantiations before doing reflection stuff - we don't want the reflection to hit the Canon anyway as those types are not directly reflectable in any form.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs:
```diff
@@ -693,7 +693,7 @@ public override void GetDependenciesDueToAccess(ref DependencyList dependencies,
 
             bool generatesMetadata = false;
-            if (!IsReflectionBlocked(writtenField))
+            if (!writtenField.OwningType.IsRuntimeDeterminedSubtype && !IsReflectionBlocked(writtenField))
```
It's usually a bug if runtime determined forms reach here. Someone should have converted to canonical. I think we might miss things if we treat these as blocked.

vitek-karas replied:
Thanks for pointing this out - fixed by passing only canonical forms from the ILScanner.

vitek-karas committed:
PR Feedback

Use Canon types for metadata analysis instead of runtime determined ones.

Added tests which hit this (for fields at least).
